Braces balance

energyc0 opened the pull request:
Braces balance exercise is done

energyc0 committed:
stack structure implementation

energyc0 committed:
ran clang-format

energyc0 committed:
Braces balance exercise

energyc0 committed:
changed strlen() to strcspn()

WoWaster reviewed (changes requested):
Не-а. PR надо направить в `main`, иначе я посмотреть на Ваш стек никак не могу. Возможно для этого придётся закрыть PR :(. 
Со вторым PR так же.

energyc0 commented:
> Не-а. PR надо направить в `main`, иначе я посмотреть на Ваш стек никак не могу. Возможно для этого придётся закрыть PR :(. Со вторым PR так же.

Направил в `main`.

energyc0 committed:
Merge branch 'main' into stackBranch

WoWaster commented on 02.10.2025/stack.c:
```diff
@@ -0,0 +1,80 @@
+    struct StackNode* prev;
+} StackNode;
+
+static StackNode* _stackNodeAlloc(int val)
```
Названия с `_` не по стайлгайду. Могу понять зачем, но "приватные" функции достаточно просто не отражать в `.h` файле. Тем более, что Вы их и так пометили `static`.

WoWaster commented on 02.10.2025/stack.c:
```diff
@@ -0,0 +1,80 @@
+static void _stackNodeFree(StackNode* node)
+{
+    free(node);
+}
```
Кажется, что этот метод почти бесполезен в текущей реализации. Лучше вызывать `free()` на месте.

WoWaster commented on 02.10.2025/stack.h:
```diff
@@ -0,0 +1,29 @@
+void stackFree(Stack* st);
+
+// Initialize stack structure
+void stackInit(Stack* st);
```
Долго не мог понять, зачем нужен этот метод. Посмотрел во вторую домашку и понял :).

WoWaster replied:
Зато не понял, зачем `stackAlloc()`, если он не используется и проще заставить пользователя вызывать `stackInit()` после *любого* выделения памяти под стек.

WoWaster commented on 02.10.2025/stack.h:
```diff
@@ -0,0 +1,29 @@
+int stackPop(Stack* st);
+
+// Return true if stack is empty
+bool stackEmpty(const Stack* st);
```
По стайлгайду: если функция возвращает `bool` хорошим тоном является использовать `is` или `has` в названии. Т.е. лучше `stackIsEmpty()`.

WoWaster commented on 02.10.2025/stack.c:
```diff
@@ -0,0 +1,80 @@
+
+int stackPeek(const Stack* st)
+{
+    return st->_data->val;
```
А если стек пуст?

WoWaster commented on 02.10.2025/stack.c:
```diff
@@ -0,0 +1,80 @@
+
+int stackPop(Stack* st)
+{
+    int data = st->_data->val;
```
Здесь такой же вопрос. Совсем умереть может быть плохо. Если пропишете в контракте функции, что пользователь должен сам всегда проверять пуст ли стек, то ок. Хотя error code был бы круче.

WoWaster commented on 02.10.2025/stack.h:
В целом, местами не хватает подробностей про контракт у функций

WoWaster commented on 02.10.2025/bracesBalance.c:
```diff
@@ -0,0 +1,67 @@
+            }
+            break;
+        default:
+            fprintf(stderr, "Undefined character.\n");
```
Правда поддержать "любые" строки сложно?)

WoWaster replied:
На самом деле содержательное, но, видимо, неактуальное, если поддержите любые строки: у Вас функция с алгоритмом внезапно орёт на пользователя -- не надо так. Вы вполне определили, что `-1` сигнализирует об ошибке. Переложите обработку ошибок на пользователя в таком случае.

WoWaster commented on 02.10.2025/bracesBalance.c:
```diff
@@ -0,0 +1,67 @@
+    if (fgets(buf, sizeof(buf), stdin) == NULL)
+        return 0;
+
+    buf[strcspn(buf, "\n")] = '\0';
```
Это, видимо, нужно, чтобы сейчас не ругалось на `\n`?

energyc0 committed:
Modified commentaries and function contract

energyc0 committed:
Merge branch 'stackBranch' into bracesBalance
Modified stack library

energyc0 committed:
Added arbitrary string support and checking its braces balance.
Return code is now 0 or 1 whether the string is balanced or not.
Do not print error message when 'non brace' character occures.

WoWaster commented on 02.10.2025/bracesBalance.c:
```diff
@@ -0,0 +1,59 @@
+ * 0    - braces are not balanced.
+ * 1    - braces are balanced.
+ */
+int checkBraceBalance(const char* str)
```
Хорошо бы в таком случае возвращать `bool`, а не `int`.

WoWaster reviewed (approved):
Комментарий по мелочи. Зачтено

energyc0 committed:
Changed checkBraceBalance return value to bool.
